### PR TITLE
🐛 Source Intercom: Fix handling of scroll param when it expired

### LINKED ISSUE #9339: 🐛 Source Intercom: `company_segment` raises 404 error while calling n+1 `companies` stream.
## Environment
- **Airbyte version**:  example is 0.22.0-alpha
- **OS Version / Instance**: example macOS, Windows 7/10, Ubuntu 18.04, GCP n2. , AWS EC2
- **Deployment**: example are Docker or Kubernetes deploy env
- **Source Connector and version**:  Intercom 0.1.12
- **Destination Connector and version**: Postgres 0.3.13
- **Severity**:Medium
- **Step where error happened**: Sync job

## Current Behavior
When syncinc `company_segments` the `companies` stream is re-consumed to build `company_segments` URL.
While performing these N+1 `companies` calls, a `scroll_param` returned by the API and used as the next page token returns 404 when called.
Could it be related to the specific behavior of scrolls in the intercom's API? 
## Expected Behavior

Can this error be handled to avoid this sync crashing?
## Logs

[logs-78-0.txt](https://github.com/airbytehq/airbyte/files/7822844/logs-78-0.txt)

## Slack conversation
https://airbytehq.slack.com/archives/C01MFR03D5W/p1641437317000700?thread_ts=1641261197.116300&cid=C01MFR03D5W

## Steps to Reproduce
The errors might be transient, other sync attempts do not have the same error.

### PR DESCRIPTION
**Pull request by lazebnyi**
## What
If the scroll isn't used for 1 minute, it expires and calls with that scroll param will fail.

## How
Working in the same way as if scroll param is already existed ([PR](https://github.com/airbytehq/airbyte/pull/8637))

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

---
This change is [Reviewable](https://reviewable.io/reviews/airbytehq/airbyte/9513)

### CODE REVIEW

**Review by annalvova05 — APPROVED** [review]

File: airbyte-integrations/connectors/source-intercom/source_intercom/source.py [hunk 1]
```diff
@@ -227,13 +228,17 @@ def should_retry(self, response: requests.Response) -> bool:
...
+                                  "Let's try to use a standard non-scroll endpoint.")
                 return False
 
             return True
         return super().should_retry(response)
 
     def backoff_time(self, response: requests.Response) -> Optional[float]:
+        if response.status_code == 404:
+            self._backoff_count = 3
+            return 0.01
```
Comment (antixar) [inline airbyte-integrations/connectors/source-intercom/source_intercom/source.py:241]: maybe will we move "3" to a separate common variable?
Can we switch to the "standard" endpoint without delays (0.01c)?
Could you add a short comment why here we set this attempt to maximum value here?

Comment (lazebnyi) [reply airbyte-integrations/connectors/source-intercom/source_intercom/source.py]: No, if backoff time will be 0. It means that we use `DefaultBackoffException`
```
custom_backoff_time = self.backoff_time(response)
            if custom_backoff_time:
                raise UserDefinedBackoffException(backoff=custom_backoff_time, request=request, response=response)
            else:
                raise DefaultBackoffException(request=request, response=response)
```

Changed scroll or standard mechanism.

File: airbyte-integrations/connectors/source-intercom/source_intercom/source.py [hunk 2]
```diff
@@ -193,6 +194,9 @@ def next_page_token(self, response: requests.Response) -> Optional[Mapping[str,
             return super().next_page_token(response)
         return None
 
+    def need_use_standard(self):
+        return not self.can_use_scroll() or self._use_standard
+
     def can_use_scroll(self):
         """Check backoff count"""
         return self._backoff_count <= 3
```
Comment (antixar) [inline airbyte-integrations/connectors/source-intercom/source_intercom/source.py:202]: can we use one of their functions? and remove another.  it seems the function `can_use_scroll` is used into the single place only and we can merge it with "need_use_standard"

Comment (lazebnyi) [reply airbyte-integrations/connectors/source-intercom/source_intercom/source.py:202]: I think better to divide these two functions. `can_use_scroll` is also used in pagination when we use scroll.

Comment (antixar) [reply airbyte-integrations/connectors/source-intercom/source_intercom/source.py:202]: Ok

**Review by antixar — APPROVED** [review]
@lazebnyi , please publish your changes

**Pull request merged**